eval/topdown.go

```diff
@@ -159,21 +194,69 @@ var builtinFunctions = map[opalog.Var]builtinFunction{
 // dereferenceVar is used to lookup the variable binding and convert the value to
 // a native Go type.
 func dereferenceVar(v opalog.Var, ctx *TopDownContext) (interface{}, error) {
-	binding, ok := ctx.Bindings[v]
-	if !ok {
+	binding := ctx.Bindings.Get(v)
+	if binding == nil {
 		return nil, fmt.Errorf("unbound variable: %v", v)
 	}
-
 	return valueToInterface(binding, ctx)
 }
 
+func evalContext(ctx *TopDownContext, iter TopDownIterator) error {
+
+	if ctx.Index >= len(ctx.Query) {
+		ctx.traceFinish()
+		return iter(ctx)
+	}
+
+	ctx.traceEval()
+
+	if ctx.Current().Negated {
+		return evalContextNegated(ctx, iter)
+	}
+
+	return evalTerms(ctx, func(ctx *TopDownContext) error {
```

Comment:
This pattern is used in a number of spots in the implementation. It's certainly possible there's a better way to implement this algorithm in Go.

The rationale was that the API should allow callers to consume query results as they're found. I.e., it's not a one shot query where all results are returned at once. There are at least three options for achieving this: (1) an iterator interface, (2) go routines, or (3) callbacks.

The first option would involve an interface type, e.g.:

```
type TopDownIterator interface {
   Value() *TopDownContext
   Next() bool
}
```

Where `Next() bool` would advance the iterator and return true if a value was available. `Value()` would return the context to process. Callers would write for loops to process results, e.g.:

```
it := someFunctionReturningIterator()
for it.Next() {
   ctx := it.Value()
   // process context
}
```

In this case, `someFunctionReturningIterator()` would return an implementation that handles some phase of the algorithm, e.g., `evalTermsIterator`.

The go routine alternative would have the different phases return channels that could be processed with a for loop, e.g.:

```
func evalTerms(...) <-chan *TopDownContext {
   ...
}

...
for ctx <- range evalTerms(...) {
  // process ctx
}
```

The callback approach is the one implemented here, using anon. functions for conciseness.

I felt like the iterator approach would be less clear because it would be using a custom iterator interface and would have more boilerplate because each phase would be a separate iterator struct + implementation.

I felt like the go routine approach may be more readable (e.g., for loops instead of callbacks) however it seems it would be harder to debug and more expensive.